lua/nvim-treesitter/node-movement.lua

```diff
@@ -0,0 +1,125 @@
+  local destination_node
+  local parent = node:parent()
+  if parent then
+    local found_pos = 0
+    for i = 0,parent:named_child_count()-1,1 do
+      if parent:named_child(i) == node then
+        found_pos = i
+        break
+      end
+    end
+    if 0 <= found_pos - 1 then
+      destination_node = parent:named_child(found_pos - 1)
+    else
+      local previous_node = get_previous_node(node:parent())
+      if previous_node and previous_node:named_child_count() > 0 then
+        destination_node = previous_node:named_child(previous_node:named_child_count() - 1)
+      elseif previous_node then
+        destination_node = previous_node
+      end
+    end
+  end
+  return destination_node
+end
+
+
+local function get_next_node(node)
+  local destination_node
+  local parent = node:parent()
+
+  if parent then
+    local found_pos = 0
+    for i = 0,parent:named_child_count()-1,1 do
+      if parent:named_child(i) == node then
+        found_pos = i
+        break
+      end
+    end
+    if parent:named_child_count() > found_pos + 1 then
+      destination_node = parent:named_child(found_pos + 1)
+    else
+      local next_node = get_next_node(node:parent())
+      if next_node and next_node:named_child_count() > 0 then
+        destination_node = next_node:named_child(0)
+      elseif next_node then
+        destination_node = next_node
+      end
+    end
+  end
+  return destination_node
+end
+
+M.do_node_movement = function(kind)
+  local buf, line, col = unpack(vim.fn.getpos("."))
+
+  local current_node = M.current_node[buf]
+
+  if current_node then
+    local node_line, node_col = current_node:start()
+    if line-1 ~= node_line or  col-1 ~= node_col then
+      current_node = nil
+    end
+  end
+  local destination_node
+
+  if parsers.has_parser() then
+    local root = parsers.get_parser():parse():root()
+    if not current_node then
+      current_node = root:named_descendant_for_range(line-1, col-1, line-1, col)
+    end
+
+
+     --UP
+    if kind == M.NodeMovementKind.up then
+       destination_node = current_node:parent()
+    -- DOWN
+    elseif kind == M.NodeMovementKind.down then
+      if current_node:named_child_count() > 0 then
+        destination_node = current_node:named_child(0)
+      end
+    -- LEFT
+    elseif kind == M.NodeMovementKind.left then
+      destination_node = get_previous_node(current_node)
+    -- RIGHT
+    elseif kind == M.NodeMovementKind.right then
+      destination_node = get_next_node(current_node)
+    end
+    M.current_node[buf] = destination_node or current_node
+  end
+
+  if destination_node then
+    node_start_to_vim(destination_node)
+  end
+end
+
+M.move_up = function() M.do_node_movement(M.NodeMovementKind.up) end
+M.move_down = function() M.do_node_movement(M.NodeMovementKind.down) end
+M.move_left = function() M.do_node_movement(M.NodeMovementKind.left) end
+M.move_right = function() M.do_node_movement(M.NodeMovementKind.right) end
+
+return M
```

Comment:
also add an `attach` and `detach` function as other modules do (you could look how its done inside `textobj`)

Reply:
so you can provide default keymaps, and it can be enabled/disabled directly from the `configs.setup` function